### ISSUE TITLE
How to locate the corresponding certificates

### ISSUE THREAD

**Issue by aghia98** [automata-network/automata-dcap-attestation#8]
Hi,
I successfully managed to run your code, specifically the "testQuoteV3OnChainAttestation()".
Now, I want to test it on my own generated quote. I guess that I should update `tcbDer`, `rootCaDer`, `platformDer`, `platformCrlDer`, `rootCrlDer` located in `PCCSSetupBase.sol` with the appropriate certificates.

I feel that I should use this Intel API to help me: https://api.portal.trustedservices.intel.com/content/documentation.html#pcs-certificate-v3
The `rootCaDer` is easilly found and corresponds to what is already hardcoded.
However, I am not sure how I should locate the other certificates. 

Should I extract some values from my quote to query them ?
Can you give me an insight about which GET request I should use ?

Thank you in advance

**Comment by smilenow** [comment]
Before using the DCAP and onchain PCCS, you need to upsert the necessary collaterals to verify your quote. We released [DCAP QPL tool](https://github.com/automata-network/automata-dcap-qpl/tree/main?tab=readme-ov-file#automata-dcap-qpl-tool) to upsert the collaterals to the target PCCS smart contracts, basically it fetches the collaterals from either Azure PCCS or Intel PCS. And you can also follow the similar approach to fetch the collaterals from:

* Your local PCCS
* Cloud providers' PCCS
* Intel PCS

It depends on how you generate the quote, should use the same data source(PCCS) to keep consistency.

Basically, in order to verify a quote, you need:

* Root CA
* Platform CA / Processor CA (it's up to which type you're using)
* Signing CA
* Root CA CRL
* Platform CA CRL / Processor CA CRL (it's up to which type you're using)
* QE Identity v3 / QE Identity v4 (it's up to which type you're using)
* the TCB info which corresponds to your FMSPC in the quote

**Comment by preston4896** [comment]
Closing this issue due to inactivity. Please re-open this issue if you had any more questions. :)

**Comment by aghia98** [comment]
Hi,
Sorry for the long time. Thank you for the provided answers, using the QPL tool would be precious to automate the verification task.
However, I want to re-open this issue since I am still trying to manually upsert the necessary collaterals in the PCCS file before moving to the automatic upsert. 
My `fmspc` is indeed different from yours, so I downloaded the corresponding `tcbinfo.json` and `QEidentity.json` from the provided Intel link. When replacing the files with the downloaded ones and applying `forge test`, I got the following error:

> [FAIL: Enclave_Id_Expired()] testQuoteV3OnChainAttestation() (gas: 1522550) 

I guess that I should still update` platformDer`, `platformCrlDer`, `rootCrlDer` (`tcbDer` and `rootCaDer` seems to be the same). But these certificates seems to be harder to get. Can you give me the steps I should follow to address my issue ?

INFO: My quote has been generated by an Azure VM

**Comment by preston4896** [comment]
Ah you probably might need to adjust the timestamp in the test setup. See https://github.com/automata-network/automata-dcap-attestation/blob/2d8b6b3dd35643081fa7fd98bbb1323a0601b2bb/forge-test/utils/PCCSSetupBase.sol#L57-L60

When running tests on Foundry with the local EVM, the block timestamp always start at 0 by default. This is why it's necessary to "warp" the EVM to a timestamp within the valid timestamp range for all of your collaterals (not just QEIdentity).

So I would suggest re-fetching all essential collaterals, and picking a timestamp that is valid for all collaterals. 

Let me know if you have any more questions.

**Comment by preston4896** [comment]
Also, have you tried verifying the quotes directly on our [testnet](https://explorer-testnet.ata.network/address/0x6D67Ae70d99A4CcE500De44628BCB4DaCfc1A145?tab=write_contract)? I think this would be an easier approach if there isn't a need to run the verifier locally.

And if the PCCS on testnet doesn't contain the collaterals needed for your quote, then it would be great to try out our QPL tool to perform upserts on the testnet.

**Comment by aghia98** [comment]
Thank you @preston4896, I will try to use the warp method with appropriate timestamp and returns to you.

The next step will be using your testnet, then I will give a try to the QPL tool

**Comment by aghia98** [comment]
Hi,
Changing the timestamp helped to overcome the previous error but I have a new one (which is pretty similar): 
`[FAIL: Certificate_Expired()] testQuoteV3OnChainAttestation() (gas: 18868013)`

(Before it was `[FAIL: Enclave_Id_Expired()] testQuoteV3OnChainAttestation() (gas: 1522550)`)

Since the consumed gas seems to be higher, I guess it is better than before but it is not working yet.

I have refetched:

- The identity.json of QE using `curl -v -X GET "https://api.trustedservices.intel.com/sgx/certification/v3/qe/identity"`. 
It starts with `{"enclaveIdentity":{"id":"QE","version":2,"issueDate":"2025-01-30T11:30:09Z","nextUpdate":"2025-03-01T11:30:09Z"...`

- The tcbinfo.json using `curl -v -X GET "https://api.trustedservices.intel.com/sgx/certification/v3/tcb?fmspc=00906ED50000` (00906ED50000 being my fmspc). 
It starts with: `{"tcbInfo":{"version":2,"issueDate":"2025-01-30T12:06:17Z","nextUpdate":"2025-03-01T12:06:17Z","fmspc":"00906ED50000"...`

The other der such as tcbDer, PlatformDer are the same as yours.
I have used: vm.warp(1738243032) which corresponds to today.

Have you any other insights ?
